subsys/net/lib/wifi_provision/wifi_provision.c

```diff
@@ -0,0 +1,995 @@
+#include <net/wifi_provision.h>
+
+#include <zephyr/net/http/client.h>
+#include <zephyr/net/http/parser.h>
+
+#include "pb_decode.h"
+#include "pb_encode.h"
+#include "proto/common.pb.h"
+
+LOG_MODULE_REGISTER(wifi_provision, CONFIG_WIFI_PROVISION_LOG_LEVEL);
+
+/* HTTP responses for demonstration */
+#define RESPONSE_200 "HTTP/1.1 200 OK\r\n"
+#define RESPONSE_400 "HTTP/1.1 400 Bad Request\r\n\r\n"
+#define RESPONSE_403 "HTTP/1.1 403 Forbidden\r\n\r\n"
+#define RESPONSE_404 "HTTP/1.1 404 Not Found\r\n\r\n"
+#define RESPONSE_405 "HTTP/1.1 405 Method Not Allowed\r\n\r\n"
+#define RESPONSE_500 "HTTP/1.1 500 Internal Server Error\r\n\r\n"
+
+/* Zephyr NET management events that this module subscribes to. */
+#define NET_MGMT_WIFI (NET_EVENT_WIFI_AP_ENABLE_RESULT		| \
+		       NET_EVENT_WIFI_AP_DISABLE_RESULT		| \
+		       NET_EVENT_WIFI_AP_STA_CONNECTED		| \
+		       NET_EVENT_WIFI_SCAN_DONE			| \
+		       NET_EVENT_WIFI_SCAN_RESULT		| \
+		       NET_EVENT_WIFI_CONNECT_RESULT		| \
+		       NET_EVENT_WIFI_AP_STA_DISCONNECTED)
+
+static struct net_mgmt_event_callback net_l2_mgmt_cb;
+
+/* Forward declarations */
+static int ap_enable(void);
+
+/* Externs */
+extern char *net_sprint_ll_addr_buf(const uint8_t *ll, uint8_t ll_len, char *buf, int buflen);
+
+/* Register service */
+DNS_SD_REGISTER_TCP_SERVICE(wifi_provision_sd, CONFIG_NET_HOSTNAME, "_http", "local",
+			    DNS_SD_EMPTY_TXT, CONFIG_WIFI_PROVISION_TCP_PORT);
+
+/* Internal variables */
+static ScanResults scan = ScanResults_init_zero;
+static uint8_t scan_result_buffer[1024];
+static size_t scan_result_buffer_len;
+static const struct smf_state state[];
+static struct http_parser_settings parser_settings;
+
+/* Semaphore used to block wifi_provision_start() until provisioning has completed. */
+static K_SEM_DEFINE(wifi_provision_sem, 0, 1);
+
+/* Variable used to indicated that we have received and stored credentials, used to break
+ * out of socket recv.
+ */
+static bool credentials_stored;
+
+/* Local reference to callers handler, used to send events to the application. */
+static wifi_provision_evt_handler_t handler_cb;
+
+/* Module events used to trigger state transitions. */
+enum module_event {
+	EVENT_AP_ENABLE = 0x1,
+	EVENT_AP_DISABLE,
+	EVENT_SCAN_DONE,
+	EVENT_CREDENTIALS_RECEIVED,
+	EVENT_RESET
+};
+
+/* Zephyr State Machine Framework variables */
+enum module_state {
+	STATE_UNPROVISIONED,
+	STATE_PROVISIONING,
+	STATE_PROVISIONED,
+	STATE_FINISHED,
+	STATE_RESET
+};
+
+/* Message queue used to schedule state transitions. */
+K_MSGQ_DEFINE(wifi_provision_msgq, sizeof(enum module_event),
+	      CONFIG_WIFI_PROVISION_MESSAGE_QUEUE_ENTRIES, 1);
+
+/* User defined state object used with SMF.
+ * Used to transfer data between state changes.
+ */
+static struct s_object {
+	/* This must be first */
+	struct smf_ctx ctx;
+
+	/* State machine variables */
+	enum module_event event_next;
+
+	/* Provisioning parameters */
+	char ssid[50];
+	char psk[50];
```

Comment:
ssid/psk have predefined maximum lengths in `wifi.h` (`WIFI_SSID_MAX_LEN`/`WIFI_PSK_MAX_LEN`), could use them instead of magic numbers.

Reply:
didnt need these, removed